pynsot/models.py

```diff
@@ -7,98 +7,595 @@
-            raise SyntaxError("Data must contain a 'data' field.")
-        obj = data['data']
-
-        if len(obj) > 1:
-            raise SyntaxError('More than one object type found.')
-        obj_type = obj.keys()[0]
-
-        if obj_type not in TYPES:
-            raise SyntaxError('Invalid type: %s' % (obj_type,))
-
-        self._object_type = obj_type
-        obj_data = obj[obj_type]
-        self.__dict__.update(obj_data)
-
-    def _repr_name(self):
-        """Normalize object type to display in repr."""
-        cls_name = self._object_type.title().replace('_', '')
-        return cls_name
+class Resource(collections.MutableMapping):
+    '''Base Resource
 
-    def __repr__(self):
-        cls_name = self._repr_name()
-        attrs = ', '.join('%s=%r' % (k, v) for k, v in self.items())
-        return '<%s(%s)>' % (cls_name, attrs)
+    This class represents a model for subclassing NSoT resources, including
+    methods to represent itself and upload to an NSoT server.
 
-    # collections.MutableMapping requires these to be defined to dictate the
-    # dictionary-like behavior.
+    Instead of overloading __init__, this base class accepts **kwargs and
+    passes them to self.postinit(**kwargs). Overload postinit for setting
+    resource specific parameters and setup. Note: This does NOT execute when
+    `raw` is provided
 
-    def __getitem__(self, val):
-        return self.__dict__[val]
+    __getitem__ and __iter__ are set so you can render what the NSoT payload
+    will look like by doing dict(obj) and fetch individual keys like a
+    dictionary with obj['attributes']
 
-    def __setitem__(self, key, val):
-        self.__dict__[key] = val
+    Attributes:
+        identifier (str)
+        resource (pynsot.vendor.slumber.Resource)
+        resource_name (str)
+        logger (logging.Logger)
+        payload (dict): This represents the exact payload sent to NSoT server
+        errors (list): Collection of any errors to happen during operations
+        last_error: The last error to happen
 
-    def __delitem__(self, key):
-        del self.__dict__[key]
+    Options:
+        site_id (int): (kwarg) Site ID for operations
+            Required unless `raw['site']`. If both provided, raw takes
+            precedence.
+        attributes (dict): (kwarg) Attributes to add to resource
+        client (pynsot.client.BaseClient): (kwarg) pynsot client
+            If client is not provided, `get_api_client` will be lazily loaded
+            on first API interaction. If you're dealing with many resource
+            objects that will talk to API, it'd be `n` less function calls to
+            pass in this parameter
+        raw (dict): (kwarg) Use this to pass in raw NSoT resource
+            Useful for instantiating objects as they return from API.
+    '''
+
+    __metaclass__ = ABCMeta
+
+    def __init__(
+        self,
+        site_id=None,
+        client=None,
+        raw={},
+        attributes={},
```

Comment:
Let's set `raw` and `attributes` to default to `None`. And then inside of `__init__()`:

```python
if raw is None:
    raw = {}
if attributes is None:
    attributes = {}
```

Because of that mutable container in function/method defs. problem.